src/equations/NavierStokes_utils.jl

```diff
@@ -0,0 +1,318 @@
+    (; x, N, dimension) = setup.grid
+    D = dimension()
+    F = ntuple(α -> similar(x[1], N), D)
+    div = similar(x[1], N)
+    p = similar(x[1], N)
+    function right_hand_side!(dudt, u, params, t)
+        u = eachslice(u; dims = ndims(u))
+        INS.apply_bc_u!(u, t, setup)
+        INS.momentum!(F, u, nothing, t, setup)
+        INS.apply_bc_u!(F, t, setup; dudt = true)
+        INS.project!(F, setup; psolver, div, p)
+        INS.apply_bc_u!(F, t, setup; dudt = true)
+        for α in 1:D
+            dudt[ntuple(Returns(:), D)..., α] .= F[α]
+        end
+        dudt
+    end
+end
+
+"""
+    pad_manual(u, D, b_axes, Iu)
+
+Manually pads the input array `u` along each dimension to match the specified axes `b_axes` using the indices `Iu`.
+
+# Arguments
+- `u::Array{T}`: The input array to be padded.
+- `D::Int`: The number of dimensions to pad.
+- `b_axes::Tuple`: A tuple containing the axes of the target array `b_INS`.
+- `Iu::NamedTuple`: A named tuple containing the indices for padding with a field `indices` which is a vector of vectors of indices. (stores in IncompressibleNavierStokes.setup)
+
+# Returns
+- `Array{T}`: The padded array.
+"""
+function pad_manual(u, D, b_axes, Iu)
+    for i in 1:D
+        x1 = (size(u, j) for j in 1:(i - 1))
+        x2 = (size(u, j) for j in (i + 1):D)
+        presize = first(Iu.indices[i]) - 1
+        postsize = last(b_axes[i]) - last(Iu.indices[i])
+        pre = similar(u, (x1..., presize, x2...))
+        post = similar(u, (x1..., postsize, x2...))
+        u = cat(pre, u, post; dims = i)
+    end
+    return u
+end
+
+"""
+    sum_NN_nopad_and_INS(a_NN, b_INS, setup)
+
+Sum a neural network output tensor `a_NN` with an incompressible Navier-Stokes tensor `b_INS` after padding `a_NN` to match the dimensions of `b_INS`.
+
+# Arguments
+- `a_NN::Array{Float32, 4}`: A tensor in the NN format, with dimensions `(n, n, dim, 1)`.
+- `b_INS::Vector{Matrix{Float32}}`: A Tuple of matrices (INS format).
+- `setup`: An IncompressibleNavierStokes.jl setup. 
+
+# Returns
+- `Tuple{Matrix{Float32}}`: A tuple of matrices (INS format) where each matrix is the sum of the padded `a_NN` tensor and the corresponding `b_INS` matrix.
+"""
+function sum_NN_nopad_and_INS(a_NN, b_INS, setup)
+    (; Iu, dimension) = setup.grid
+    D = dimension()
+    griddims = ((:) for _ in 1:D)
+    b_INS_axes = Tuple(axes(b_INS[i]) for i in eachindex(b_INS))
+    Tuple(
+        pad_manual(a_NN[griddims..., i, 1], D, b_INS_axes[i], Iu[i]) + b_INS[i]
+    for i in 1:length(b_INS)
+    )
+end
+
+"""
+    INS_to_NN(u, setup)
+
+Converts the input velocity field `u` from the IncompressibleNavierStokes.jl style `u[time step]=(ux, uy)`
+to a format suitable for neural network training `u[n, n, D, batch]`.
+
+# Arguments
+- `u`: Velocity field in INS style.
+
+# Returns
+- `u`: Velocity field converted to a tensor format suitable for neural network training.
+"""
+function INS_to_NN(u)
+    cat(u...; dims = ndims(u[1]) + 1)
```

Comment:
why not this to avoid copy and allocation
```suggestion
    view(u, 1)
```

Reply:
@v1kko do you have an opinion about this?

Reply:
It looks good, I say use it if it works

Reply:
No, it doesn't work. Indeed u is a Tuple, and there is now `view` function typed for `Tuple`. Here the code that you can try to see that it doesn't work:
```julia
u0 = INS.random_field(setups[1], 0)
op1 = cat(u0...; dims = ndims(u0[1]) + 1) # current, working
op2 = view(u0, 1) # proposed, doesn't work
```